File: sgd.py
import numpy as np
from matplotlib import pyplot
import pandas
from pandas import HDFStore


def sgd(R,Theta,X,K,iterations = 1000000, alpha=0.0005, beta=0.02):
    errors = []
    ploterrors = []
    X = X.T # Transpose movie feature matrix to be conformable in dot products
    for iter in range(iterations):
        if (iter+1)%100000 == 0:
            print("Starting iteration", iter+1)
            emean = np.mean(errors[iter-99999:]) # Hack to avoid indexing error at first 100k checkpoint
            print("Moving average error:", emean)
            ploterrors.append(emean)
        try:
            e = 0
            rnd = int(np.round((len(R)-1)*np.random.rand()))  # Select random user
            i = R['userId'][rnd]
            j = R['movieId'][rnd]
            fasit = R['rating'][rnd]

            eij = fasit - Theta[i,:].dot(X[:,j])  # Compute prediction error for sample
            Theta[i,:] +=  alpha*(2*eij*X[:,j] - beta*Theta[i,:])  # Adjust user features
            X[:,j] +=  alpha*(2*eij*Theta[i,:] - beta*X[:,j])  # Adjust movie features

            e += pow(fasit - Theta[i,:].dot(X[:,j]), 2)  # Compute squared error for plotting purposes
            errors.append(e)
        except MemoryError:
            print("memory error")

    return Theta, X.T, ploterrors


def learn(iterations=1000000):
    print("Loading HDF5 file...")
    store = HDFStore('ratings.h5')
    R = store['r100k']

    N = max(R['userId'])
    M = max(R['movieId'])
    K = 20

    print("Generating user features...")
    Theta = np.random.rand(N+1, K)
    print("Generating movie features...")
    X = np.random.rand(M+1, K)

    print("Starting Stochastic Gradient Descent...")
    nTheta, nX, errors = sgd(R, Theta, X, K, iterations)

    pyplot.figure(figsize=(10, 6))
    pyplot.plot(errors)
    pyplot.title("Loss function")
    pyplot.grid(True)
    pyplot.show()

    return nTheta, nX, errors


def predict(user, movie, Theta, X):
    X = X.T
    rating = Theta[user,:].dot(X[:,movie])
    print("Prediction for user", user, "movie", movie, ":", rating)
    return rating


def compute_user_averages(R):
    num = max(R['userId'])+1
    avg_user_ratings = [0]*num
    for i in range(num):
        avg_user_ratings[i] = (np.mean(R['rating'][R['userId']==i]))
    return avg_user_ratings


def compute_movie_averages(R):
    num = max(R['movieId'])+1
    avg_movie_ratings = [0]*num
    for i in range(num):
        avg_movie_ratings[i] = (np.mean(R['rating'][R['movieId']==i]))
    return avg_movie_ratings


def remove_nans(list):
    for i in range(len(list)):
        if(str(list[i])=='nan'):
            list[i] = 0

def normalize(R, movie_avgs, usereffects):
    pandas.options.mode.chained_assignment = None # Allow assignment back to R
    normalized = []
    for i in range(len(R)):
        R['rating'][i] -= (movie_avgs[0][R['movieId'][i]] + usereffects[0][R['userId'][i]])


def predict_normalized(user, movie, Theta, X, movie_avg, usereffect):
    X = X.T
    rating = movie_avg + usereffect + Theta[user,:].dot(X[:,movie])
    print("Prediction for user", user, "movie", movie, ":", rating)
    return rating

File: sgd.jl
using PyPlot
using DataFrames
using HDF5

function SGD(R,Theta,X,K,epochs=1000000,alpha=0.0002,beta=0.02)
    errors = []
    ploterrors = []
    X = transpose(X)
    for iter = 1:epochs
        if iter%10000 == 0
            println("Starting iteration ", iter+1)
            emean = mean(errors[end-9998:end])
            println("Moving average error: ", emean)
            push!(ploterrors, emean)
        end
        e = 0
        rnd = rand(1:maximum(R[1]))
        i = R[2][rnd]
        j = R[3][rnd]
        realrating = R[4][rnd]

        eij = realrating - dot(vec(Theta[i,:]),vec(X[:,j]))
        Theta[i,:] = vec(Theta[i,:]) + alpha*(2*eij*vec(X[:,j]) - beta*vec(Theta[i,:]))
        X[:,j] = vec(X[:,j]) + alpha*(2*eij*vec(Theta[i,:]) - beta*vec(X[:,j]))

        e += (realrating - dot(vec(Theta[i,:]),vec(X[:,j])))^2
        push!(errors,e)
    end

    return Theta, transpose(X), ploterrors
end


println("Loading HDF5 file...")
r = h5read("c://Dev//Code//movie-recommender//ratings.h5", "r100k")
println("Loading to data frame...")
R = DataFrame(index=r["axis1"]+1,userId=vec(r["block1_values"][1,:]),movieId=vec(r["block1_values"][2,:]),rating=vec(r["block0_values"]),timestamp=vec(r["block1_values"][3,:]))

N = maximum(R[2])
M = maximum(R[3])
K = 10

println("Generating random user features...")
Theta = rand(N,K)
println("Generating random movie features...")
X = rand(M,K)

println("Starting Stochastic Gradient Descent...")

tic()
nTheta, nX, errors = SGD(R,Theta,X,K)
toc()

nR = nTheta*transpose(nX)

# error = 0
# count = 0
# for i in 1:10
#     for j in 1:10
#         if R[i,j] > 0
#             error += (R[i,j] - nR[i,j])^2
#             count += 1
#         end
#     end
# end
# println("Mean squared error: ", error/count)

predict = dot(vec(nTheta[1,:]),vec(transpose(nX)[:,16]))

plot(errors)
